what4/src/What4/SpecialFunctions.hs

```diff
@@ -0,0 +1,285 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+{-|
+Module           : What4.SpecialFunctions
+Description      : Utilities relating to special functions
+Copyright        : (c) Galois, Inc 2021
+License          : BSD3
+Maintainer       : Rob Dockins <rdockins@galois.com>
+
+Utilties for representing and handling certain \"special\"
+functions arising from analysis. Although many of these
+functions are most properly understood as complex valued
+functions on complex arguments, here we are primarily interested
+in their restriction to real-valued functions or their
+floating-point approximations.
+
+The functions considered here include functions from
+standard and hyperbolic trigonometry, exponential
+and logarithmic functions, etc.  Some of these functions
+are defineable in terms of others (e.g. @tan(x) = sin(x)/cos(x)@
+or expm1(x) = exp(x) - 1@) but are commonly implemented
+separately in math libraries for increased precision.
+Some popular constant values are also included.
+-}
+
+module What4.SpecialFunctions
+  ( R
+  , SpecialFunction(..)
+  , buildSpecialFnArgs
+  , FunctionSymmetry(..)
+  , specialFnSymmetry
+  , SpecialFnArg(..)
+  , traverseSpecialFnArg
+  , SpecialFnArgs(..)
+  , traverseSpecialFnArgs
+  ) where
+
+import           Data.Kind (Type)
+import           Data.Parameterized.Classes
+import qualified Data.Parameterized.Context as Ctx
+import           Data.Parameterized.Context ( pattern (:>) )
+import           Data.Parameterized.Ctx
+import           Data.Parameterized.TH.GADT
+import           Data.Parameterized.TraversableFC
+
+data FunctionSymmetry r
+  = NoSymmetry
+  | EvenFunction
+  | OddFunction
+ deriving (Show)
+
+data R
+
+-- | Data type for representing
```

Comment:
Is "Data type for representing" a cut-off sentence?

I was sort of wondering what the criteria was for something to qualify as a special function (e.g., does it suffice to be mentioned in the "Recommended operations" section of the [IEEE 754 standard](https://irem.univ-reunion.fr/IMG/pdf/ieee-754-2008.pdf)?), which might be useful to list here.

Reply:
Indeed, I forgot to finish this sentence.

The "recommended operations" list is probably a good place to start, but I imagine we'll add to this collection when and as needed.  On awkward thing I notice about that list is that some of the operation take integer arguments, whereas so far I've only involved real arguments here.